File: packages/flutter-im-ui/lib/src/models/rich_doc.dart
/// A rich-text message as it is drawn: the RichDoc v2 document the core stores
/// (`docJson`, flare-proto `RichTextContent.doc_json`) read into blocks and
/// runs. The rule is shared with the other three kits and written down in
/// `spec/rich-doc-vectors.json`; Vue implements it in `utils/richDoc.ts`.
library;

import 'dart:convert';

import 'url_safety.dart';

/// The marks a run can carry, in the order a run lists them.
enum FlareRichMark { bold, italic, underline, strike, spoiler }

/// One run of a paragraph or heading.
class FlareRichRun {
  const FlareRichRun(
    this.text, {
    this.marks = const [],
    this.code = false,
    this.link,
    this.mention,
    this.emoji,
  });

  final String text;

  /// The known marks, once each, in [FlareRichMark] order.
  final List<FlareRichMark> marks;

  /// Inline code.
  final bool code;

  /// The href of the innermost link around the run that [safeExternalUrl]
  /// accepts, as written.
  final String? link;

  /// A mention: the user id it names, possibly empty.
  final String? mention;

  /// An emoji: its pack key, possibly empty.
  final String? emoji;

  bool has(FlareRichMark mark) => marks.contains(mark);

  FlareRichRun _linked(String? href) => href == null
      ? this
      : FlareRichRun(
          text,
          marks: marks,
          code: code,
          link: href,
          mention: mention,
          emoji: emoji,
        );
}

/// A block of a rich-text body.
sealed class FlareRichBlock {
  const FlareRichBlock();
}

class FlareRichParagraph extends FlareRichBlock {
  const FlareRichParagraph(this.runs);
  final List<FlareRichRun> runs;
}

class FlareRichHeading extends FlareRichBlock {
  const FlareRichHeading(this.level, this.runs);

  /// 1…6.
  final int level;
  final List<FlareRichRun> runs;
}

class FlareRichQuote extends FlareRichBlock {
  const FlareRichQuote(this.blocks);
  final List<FlareRichBlock> blocks;
}

class FlareRichCode extends FlareRichBlock {
  const FlareRichCode(this.text, {this.language});
  final String text;
  final String? language;
}

class FlareRichList extends FlareRichBlock {
  const FlareRichList({required this.ordered, required this.items});
  final bool ordered;

  /// Each item's blocks; never empty.
  final List<List<FlareRichBlock>> items;
}

class FlareRichDivider extends FlareRichBlock {
  const FlareRichDivider();
}

/// What a covered spoiler run draws in place of [text]: every code point that
/// is not whitespace becomes U+3000 IDEOGRAPHIC SPACE, so the words are in
/// neither the rendered text nor a copy until the reader reveals them.
String flareRichSpoilerCover(String text) => String.fromCharCodes(
  text.runes.map(
    (rune) => _whitespace.hasMatch(String.fromCharCode(rune)) ? rune : 0x3000,
  ),
);

final _whitespace = RegExp(r'\s');

/// The core's own limits (`rich_doc_v2/validate.rs`).
const int flareRichDocMaxDepth = 64;
const int flareRichDocMaxNodes = 10000;

/// The blocks a rich-text body draws for [input] — the document's JSON text
/// or its decoded map — or null when it is not a drawable RichDoc v2 document,
/// and the body shows the message's plain text instead.
List<FlareRichBlock>? flareParseRichDoc(Object? input) {
  Object? value = input;
  if (input is String) {
    try {
      value = jsonDecode(input);
    } on FormatException {
      return null;
    }
  }
  if (value is! Map) return null;
  final version = value['version'];
  if (value['type'] != 'doc' ||
      version is! num ||
      version != 2 ||
      value['children'] is! List) {
    return null;
  }
  if (!_withinLimits(value)) return null;
  return _blocks(value['children'] as List);
}

List<Object?> _children(Object? node) => node is Map && node['children'] is List
    ? node['children'] as List
    : const [];

String _string(Object? value) => value is String ? value : '';

/// Whether the tree stays inside the core's depth and node limits.
bool _withinLimits(Map root) {
  var nodes = 1;
  final stack = <(Object?, int)>[
    for (final child in _children(root)) (child, 1),
  ];
  while (stack.isNotEmpty) {
    final (node, depth) = stack.removeLast();
    if (depth > flareRichDocMaxDepth) return false;
    nodes += 1;
    if (nodes > flareRichDocMaxNodes) return false;
    for (final child in _children(node)) {
      stack.add((child, depth + 1));
    }
  }
  return true;
}

List<FlareRichRun> _runs(
  List<Object?> inlines,
  String? link, [
  List<FlareRichRun>? into,
]) {
  final out = into ?? <FlareRichRun>[];
  void push(FlareRichRun run) => out.add(run._linked(link));
  for (final node in inlines) {
    if (node is! Map) continue;
    switch (node['type']) {
      case 'text':
        final text = _string(node['text']);
        if (text.isEmpty) break;
        final named = <Object?>{
          for (final mark
              in node['marks'] is List ? node['marks'] as List : const [])
            if (mark is Map) mark['type'],
        };
        push(
          FlareRichRun(
            text,
            marks: [
              for (final mark in FlareRichMark.values)
                if (named.contains(mark.name)) mark,
            ],
          ),
        );
      case 'inline_code':
        final text = _string(node['text']);
        if (text.isNotEmpty) push(FlareRichRun(text, code: true));
      case 'hard_break':
        push(const FlareRichRun('\n'));
      case 'mention':
        final userId = _string(node['user_id']);
        final text = _string(node['text']);
        if (userId.isNotEmpty || text.isNotEmpty) {
          push(
            FlareRichRun(text.isNotEmpty ? text : '@$userId', mention: userId),
          );
        }
      case 'emoji':
        final key = _string(node['key']);
        final text = _string(node['text']);
        if (key.isNotEmpty || text.isNotEmpty) {
          push(FlareRichRun(text.isNotEmpty ? text : ':$key:', emoji: key));
        }
      case 'link':
        final href = _string(node['href']);
        final accepted = href.isNotEmpty && safeExternalUrl(href) != null;
        _runs(_children(node), accepted ? href : link, out);
      case 'custom_inline':
        break;
      default:
        final text = _string(node['text']);
        if (text.isNotEmpty) push(FlareRichRun(text));
    }
  }
  return out;
}

List<FlareRichBlock> _blocks(
  List<Object?> values, [
  List<FlareRichBlock>? into,
]) {
  final out = into ?? <FlareRichBlock>[];
  for (final node in values) {
    if (node is! Map) continue;
    switch (node['type']) {
      case 'paragraph':
        final runs = _runs(_children(node), null);
        if (runs.isNotEmpty) out.add(FlareRichParagraph(runs));
      case 'heading':
        final runs = _runs(_children(node), null);
        if (runs.isEmpty) break;
        final level = node['level'];
        final raw = level is num && level.isFinite ? level.truncate() : 1;
        out.add(FlareRichHeading(raw.clamp(1, 6), runs));
      case 'quote':
        final blocks = _blocks(_children(node));
        if (blocks.isNotEmpty) out.add(FlareRichQuote(blocks));
      case 'code_block':
        final text = _children(node).map((child) {
          if (child is! Map) return '';
          if (child['type'] == 'text') return _string(child['text']);
          return child['type'] == 'hard_break' ? '\n' : '';
        }).join();
        if (text.isEmpty) break;
        final language = _string(node['language']);
        out.add(
          FlareRichCode(text, language: language.isEmpty ? null : language),
        );
      case 'bullet_list':
      case 'ordered_list':
        final items = [
          for (final child in _children(node))
            if (child is Map && child['type'] == 'list_item')
              _blocks(_children(child)),
        ].where((item) => item.isNotEmpty).toList();
        if (items.isNotEmpty) {
          out.add(
            FlareRichList(
              ordered: node['type'] == 'ordered_list',
              items: items,
            ),
          );
        }
      case 'divider':
        out.add(const FlareRichDivider());
      case 'custom_block':
        _blocks(_children(node), out);
      default:
        break;
    }
  }
  return out;
}

File: packages/android-im-ui/src/main/kotlin/com/flare/im/ui/RichDoc.kt
package com.flare.im.ui

/*
 * A rich-text message as it is drawn: the RichDoc v2 document the core stores (`docJson`, flare-proto
 * `RichTextContent.doc_json`) read into blocks and runs. The rule is shared with the other three kits and
 * written down in `spec/rich-doc-vectors.json`; Vue implements it in `utils/richDoc.ts`.
 */

/** The marks a run can carry, in the order a run lists them. */
enum class FlareRichMark(val wire: String) {
    Bold("bold"), Italic("italic"), Underline("underline"), Strike("strike"), Spoiler("spoiler"),
}

/**
 * One run of a paragraph or heading. [marks] are the known marks once each, in [FlareRichMark] order; [link] is
 * the href of the innermost link around the run that [safeExternalUrl] accepts, as written; [mention] is the
 * user id a mention names and [emoji] an emoji's pack key (either possibly empty).
 */
data class FlareRichRun(
    val text: String,
    val marks: List<FlareRichMark> = emptyList(),
    val code: Boolean = false,
    val link: String? = null,
    val mention: String? = null,
    val emoji: String? = null,
) {
    fun has(mark: FlareRichMark) = mark in marks
}

/** A block of a rich-text body. */
sealed interface FlareRichBlock {
    data class Paragraph(val runs: List<FlareRichRun>) : FlareRichBlock
    /** [level] is 1…6. */
    data class Heading(val level: Int, val runs: List<FlareRichRun>) : FlareRichBlock
    data class Quote(val blocks: List<FlareRichBlock>) : FlareRichBlock
    data class Code(val text: String, val language: String? = null) : FlareRichBlock
    /** Each item's blocks; never empty. */
    data class ListBlock(val ordered: Boolean, val items: List<List<FlareRichBlock>>) : FlareRichBlock
    data object Divider : FlareRichBlock
}

/**
 * What a covered spoiler run draws in place of [text]: every code point that is not whitespace becomes U+3000
 * IDEOGRAPHIC SPACE, so the words are in neither the rendered text nor a copy until the reader reveals them.
 */
fun flareRichSpoilerCover(text: String): String = buildString {
    text.codePoints().forEach { codePoint -> if (Character.isWhitespace(codePoint)) appendCodePoint(codePoint) else append('\u3000') }
}

/** The core's own limits (`rich_doc_v2/validate.rs`). */
const val FLARE_RICH_DOC_MAX_DEPTH = 64
const val FLARE_RICH_DOC_MAX_NODES = 10_000

/**
 * The blocks a rich-text body draws for [input] — the document's JSON text or its decoded map — or null when it
 * is not a drawable RichDoc v2 document, and the body shows the message's plain text instead.
 */
fun flareParseRichDoc(input: Any?): List<FlareRichBlock>? {
    val value = if (input is String) {
        try { FlareJson.parse(input) } catch (_: IllegalArgumentException) { return null } catch (_: NumberFormatException) { return null }
    } else {
        input
    }
    val root = value as? Map<*, *> ?: return null
    val version = root["version"]
    if (root["type"] != "doc" || version !is Number || version.toDouble() != 2.0 || root["children"] !is List<*>) return null
    val children = root["children"] as List<*>
    if (!withinLimits(children)) return null
    return blocks(children)
}

private fun childrenOf(node: Any?): List<*> = ((node as? Map<*, *>)?.get("children") as? List<*>).orEmpty()

private fun stringOf(value: Any?): String = value as? String ?: ""

private fun withinLimits(top: List<*>): Boolean {
    var nodes = 1
    val stack = ArrayDeque<Pair<Any?, Int>>()
    top.forEach { stack.addLast(it to 1) }
    while (stack.isNotEmpty()) {
        val (node, depth) = stack.removeLast()
        if (depth > FLARE_RICH_DOC_MAX_DEPTH) return false
        nodes += 1
        if (nodes > FLARE_RICH_DOC_MAX_NODES) return false
        childrenOf(node).forEach { stack.addLast(it to depth + 1) }
    }
    return true
}

private fun runs(inlines: List<*>, link: String?, out: MutableList<FlareRichRun> = mutableListOf()): MutableList<FlareRichRun> {
    fun push(run: FlareRichRun) { out += if (link == null) run else run.copy(link = link) }
    for (value in inlines) {
        val node = value as? Map<*, *> ?: continue
        when (node["type"]) {
            "text" -> {
                val text = stringOf(node["text"])
                if (text.isEmpty()) continue
                val named = (node["marks"] as? List<*>).orEmpty().mapNotNull { (it as? Map<*, *>)?.get("type") }.toSet()
                push(FlareRichRun(text, marks = FlareRichMark.entries.filter { it.wire in named }))
            }
            "inline_code" -> stringOf(node["text"]).takeIf { it.isNotEmpty() }?.let { push(FlareRichRun(it, code = true)) }
            "hard_break" -> push(FlareRichRun("\n"))
            "mention" -> {
                val userId = stringOf(node["user_id"])
                val text = stringOf(node["text"])
                if (userId.isNotEmpty() || text.isNotEmpty()) push(FlareRichRun(text.ifEmpty { "@$userId" }, mention = userId))
            }
            "emoji" -> {
                val key = stringOf(node["key"])
                val text = stringOf(node["text"])
                if (key.isNotEmpty() || text.isNotEmpty()) push(FlareRichRun(text.ifEmpty { ":$key:" }, emoji = key))
            }
            "link" -> {
                val href = stringOf(node["href"])
                val accepted = href.isNotEmpty() && safeExternalUrl(href) != null
                runs(childrenOf(node), if (accepted) href else link, out)
            }
            "custom_inline" -> Unit
            else -> stringOf(node["text"]).takeIf { it.isNotEmpty() }?.let { push(FlareRichRun(it)) }
        }
    }
    return out
}

private fun blocks(values: List<*>, out: MutableList<FlareRichBlock> = mutableListOf()): MutableList<FlareRichBlock> {
    for (value in values) {
        val node = value as? Map<*, *> ?: continue
        when (val type = node["type"]) {
            "paragraph" -> runs(childrenOf(node), null).takeIf { it.isNotEmpty() }?.let { out += FlareRichBlock.Paragraph(it) }
            "heading" -> {
                val inline = runs(childrenOf(node), null)
                if (inline.isEmpty()) continue
                val raw = (node["level"] as? Number)?.toDouble()?.takeIf { it.isFinite() }?.toInt() ?: 1
                out += FlareRichBlock.Heading(raw.coerceIn(1, 6), inline)
            }
            "quote" -> blocks(childrenOf(node)).takeIf { it.isNotEmpty() }?.let { out += FlareRichBlock.Quote(it) }
            "code_block" -> {
                val text = childrenOf(node).joinToString("") { child ->
                    val inline = child as? Map<*, *>
                    when (inline?.get("type")) {
                        "text" -> stringOf(inline["text"])
                        "hard_break" -> "\n"
                        else -> ""
                    }
                }
                if (text.isEmpty()) continue
                out += FlareRichBlock.Code(text, stringOf(node["language"]).ifEmpty { null })
            }
            "bullet_list", "ordered_list" -> {
                val items = childrenOf(node)
                    .filter { (it as? Map<*, *>)?.get("type") == "list_item" }
                    .map { blocks(childrenOf(it)) }
                    .filter { it.isNotEmpty() }
                if (items.isNotEmpty()) out += FlareRichBlock.ListBlock(ordered = type == "ordered_list", items = items)
            }
            "divider" -> out += FlareRichBlock.Divider
            "custom_block" -> blocks(childrenOf(node), out)
            else -> Unit
        }
    }
    return out
}
